Extract common GetData logic from framework-specific AutoDataAttribute classes

olegsych opened the pull request:
Defining test data customization attributes that are specific to a _mocking_ framework without taking a dependency on a specific _test_ framework is not possible today because the `CustomizeAttribute` classes are defined in the _test framework assemblies_, like `Ploeh.AutoFixture.Nunit2`, `Ploeh.AutoFixture.Xunit` and `Ploeh.AutoFixture.Xunit2`, while the mocking attributes, such as the `SubstituteAttribute` added in #430, have to be defined in the _mocking framework assemblies_, like the `Ploeh.AutoFixture.NSubstitute`. Because of this limitation, the usage of the `SubstituteAttribute` can produce obscure test failures, described in #430, due to missing fixture customizations.

Here is the desired usage scenario:

```C#
using Ploeh.AutoFixture.NSubstitute;
using Ploeh.AutoFixture.xUnit;
using Xunit;

public class MyTest
{
    [Fact, AutoData]
    public void Test1([Substitute]MyType sample)
    {
        // Fixture setup
        // Exercise system
        sample.HelloWorld();
        // Verify results
        sample.Received().HelloWorld();
        // Teardown
    }
}
```

Today, this test fails with an obscure error saying that the object instance for which the `Received` method is called is not a substitute. Developer writing this test has to know that an `AutoNSubstituteCustomization` or an `AutoConfiguredNSubstituteCustomization` has to be used for this test to work correctly.

To solve this problem, this pull request 
- introduces a new Kernel interface, `IParameterCustomizationProvider` that can be implemented by attributes defined by any of the _mocking_ assemblies, like `AutoFixture.NSubstitute` and `AutoFixture.AutoMoq`, without additional dependencies. 
- implements a new `TestDataProvider` utility class that uses `IParameterCustomizationProvider` and consolidates the fixture customization logic from the three separate `AutoDataAttribute` implementations in `AutoFixture.Nunit2`, `AutoFixture.xUnit` and `AutoFixture.xUnit2`. 
- modifies the existing `AutoDataAttribute` implementationsto use the common logic in the `TestDataProvider`.

Once this pull request is merged, the existing `SubstituteAttribute` can be changed to implement the new `IParameterCustomizationProvider` interface, which will make the test in the scenario above work as expected.

ploeh commented:
Thank you for your interest in AutoFixture.

AFAICT, this contribution creates new public types that we will have to maintain for a long time. There ought to be some benefits that outweigh that disadvantage; what are they?

This is reminiscent of [a discussion we've already had](https://github.com/AutoFixture/AutoFixture/issues/118). Additionally, I'd like to point out Udi Dahan's excellent article [Beware the Share](http://programmer.97things.oreilly.com/wiki/index.php/Beware_the_Share).

olegsych commented:
Thanks for responding so quickly, @ploeh. I know you're busy, so let me try to answer your concerns as completely and succinctly as I can. I will also update the pull request description to include a better description of the scenarios this pull request enables.

> AFAICT, this contribution creates new public types that we will have to maintain for a long time. There ought to be some benefits that outweigh that disadvantage; what are they?

- The `IParameterCustomizationProvider` interface enables development of mocking framework-specific attributes _independently_ of the test framework-specific attributes. 
- Implementing an attribute like `[Substitute]` discussed in #132 without the `IParameterCustomizationProvider` interface, would require breaking the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) to support all three test frameworks currently supported by AutoFixture.
- The `TestDataProvider` consolidates 3 separate, identical implementation of GetData methods from `AutoFixture.Nunit2`, `AutoFixture.xUnit.net` and `AutoFixture.xUnit.net2`, reducing the future maintenance cost.
- The `TestDataProviderTest` provides comprehensive test coverage for the previously _untested_ logic of the `AutoDataAttribute` implementations that apply `ICustomization` instances returned by `CustomizeAttribute` classes to the test `IFixture`. 
- Implementing support for the `IParameterCustomizationProvider` interface in the `AutoDataAttribute` classes in the test framework assemblies, without introduction of the `TestDataProvider` would break the _rule of three_.

> This is reminiscent of a discussion we've already had. Additionally, I'd like to point out Udi Dahan's excellent article Beware the Share.

-  Since #118 was closed 2 years ago, the introduction of support for XUnit v2 has already broke the _rule of three_ :smile:. AutoFixture now has 3 separate copies of the `AutoDataAttribute` and `CustomizeAttribute` classes. Taking into consideration that the `CustomizeAttribute` seems to have withstood the test of time and reuse, I think it's worthwhile to consider reopening this discussion.
- While consolidating separate `CustomizeAttribute` definitions into a single class in a separate NuGet package referenced by the test framework packages would be ideal, the proposed solution in this pull request makes a step toward that solution without breaking backward compatibility for existing consumers of v3.x. 
- A separate assembly to hold the consolidated `CustomizeAttribute` could also be introduced without breaking backward compatibility with the help of [TypeForwardedToAttribute](https://msdn.microsoft.com/en-us/library/system.runtime.compilerservices.typeforwardedtoattribute(v=vs.110).aspx) in the test framework assemblies. The advantage of doing it this way is that it allows eliminating the `IParameterCustomizationProvider` interface at the cost of adding a new project/assembly/package to AutoFixture. Depending on how many other types can be immediately moved there, this could be a better alternative.
- While I completely agree with Udi's points regarding the cost of sharing code, in this particular case, _not sharing_ the `GetData` implementation creates a bigger problem because excessive duplication can affects the code similarly to the over-aggressive sharing does. Because of the reasons listed above, evolution of the mocking framework assemblies, such as `AutoFixture.NSubstitute`, independently from test framework assemblies, such as `AutoFixture.NUnit2` is effectively discouraged today. The current implementation of `AutoDataAttribute` classes encourages additional duplication and dependencies because requires creation of assemblies, like `AutoFixture.NSubstitute.Nunit2`, `AutoFixture.NSubstitute.xUnit.net` and `AutoFixture.NSubstitute.xUnit.net2` and so on, while discouraging thorough testing due to the geometric rate of code growth.

ploeh commented:
I'm torn.

There are many good points here. OTOH, I'm beginning to think that the AutoFixture core library is becoming quite bloated, and I'm loath to put even more stuff into it.

From my perspective, the Rule of Three describes a *minimum threshold:* before one has the same code in triplicate, there's too little information to separate the differences from the similarities. IME, this doesn't meant that one *must* refactor to remove the duplication as soon as the duplication count reaches three, but rather that this refactoring shouldn't be done *before* that threshold.

With the imminent release of *.NET Core*, it looks like there's much work ahead getting AutoFixture working in other configurations (re: #404). I don't think it'd be possible to add .NET Core support to AutoFixture without breaking changes, so we're looking at a potential AutoFixture 4.

The reason I bring this up is that this will most likely involve removing a lot of code (and features) from the core AutoFixture library, so I'm not sure now is the best time to *add* code.

As initially stated, I'm in doubt, so I'd like to solicit other opinions; particularly those of the other core contributors @moodmosaic, @ecampidoglio, and @adamchester.

moodmosaic commented:
This looks quite impressive piece of work :smiley:

I think it goes off on a tangent, as the core AutoFixture library isn't aware of test methods and parameters – that's where AutoFixture with xUnit.net/NUnit Glue Libraries come into play, optionally followed by one of the auto-mocking Glue Libraries.

After all, it becomes quite common scenario to compose multiple Customizations, so taking a dependency on a specific test framework shouldn't be a problem per se.

That being said, I can imagine users of the core AutoFixture library being quite confused by the proposed `IParameterCustomizationProvider`.

(On a side note: I haven't gone through all the changes, but I'm wondering whether some implementations of `IParameterCustomizationProvider` can be subject to the [Constrained Construction anti-pattern](http://blog.ploeh.dk/2011/04/27/Providerisnotapattern/).)

IMHO, taking into account @ploeh's [comment](https://github.com/AutoFixture/AutoFixture/pull/418#issuecomment-120075929) and given all the above, I wouldn't move forward with this...

olegsych commented:
Now that the `SubstituteAttribute` is implemented, I'd like to continue working on this pull request because it provides the foundation required to prevent the obscure test failures due to missing NSubstitute customization described in #430.

ploeh commented:
I must admit that I forgot all about this pull request :flushed: 

Does it add anything that AutoFixture *needs?* In other words, are there features we can't add to AutoFixture, that we could add with the code in this pull request?

olegsych commented:
I have updated description of the pull request to answer these questions and accurately reflect the current implementation of `AutoFixture`.

ploeh commented:
The design philosophy behind AutoFixture is that it tries to follow the 80-20 rule. If it can understand 80 % of your API, it's already done a lot of work for you, but given that a user could do *anything* with their API, it'd be impossible to address all quirks.

AutoFixture is quite opinionated in how it expects an API to be designed. Those opinions are based on fundamental principles of OOD, such as *design by contract*.

Even though I'm the main designer of AutoFixture, and the opinions embedded in it reflect mine, I rarely use AutoFixture without tweaking it.

First, you may find yourself in a situation where you want to write a test with this signature:

```C#
[Fact, AutoData]
public void Test1(MyType sample, IMyInterface iface)
```

That's not going to be possible, because vanilla AutoFixture doesn't know how to create an instance of an interface. You'll have to create a custom attribute, and pull in one or more customizations, such as `AutoNSubstituteCustomization`.

Yes, you'll have to know how to do that, but in the motivating example here, you'll also have to know that a `[Substitute]` attribute exists.

Even so, it rarely stops there. Often (in practice: *always*) you'll have to encapsulate some domain rule that you'll have to tell AutoFixture about. Imagine, for example, that you'll have to deal with [ISBNs](https://en.wikipedia.org/wiki/International_Standard_Book_Number) in your code. You may introduce a Value Object that encapsulates the rules about ISBN (valid digits, check sums), but you still need to populate the Value Object with values. You may, for example, pass a `string` to the constructor, but AutoFixture's default string algorithm isn't going to produce valid ISBNs. This means that you have to customise how AutoFixture generates ISBNs.

The point here is that no matter how much you try, you can't eliminate the need for customising AutoFixture. Thus, I'm not particularly concerned about the need for explicitly pulling in `AutoNSubstituteCustomization`.

Another design principle behind AutoFixture is from the [Zen of Python](https://www.python.org/dev/peps/pep-0020) - particularly the second rule: *Explicit is better than implicit*.

For that reason, I don't wish to pull AutoFixture in a direction where it's becoming even more 'auto-magical' than it already is.

In closing, I don't think this proposed change solves a real problem, and while I think you for your interest, I don't want to inflate the kernel's public API without a sound reason.

olegsych commented:
Thanks for sharing your thoughts on this. I do appreciate it. I hope I can trouble you with one last question. :fearful:   

Could you help me understand why you are rejecting a new implementation of what appears to be an already established AutoFixture design pattern? Using examples from you previous comment, you don't want to allow the following usage of AutoFixture.

```C#
[Theory, AutoData]
public void Test1([Substitute]IMyInterface iface)
```

While the following usage is allowed and already works _as is_, no extra code needed.

```C#
[Theory, AutoData]
public void Test1([RegularExpression(@"\d{9}[\d|X]")]string isbn)
```

As a user of AutoFixture looking at this hypothetical code written by someone else on my team, I'd be surprised that the `System.ComponentModel.DataAnnotations` attributes work out of the box even though they have no relation to AutoFixture, while a built-in attribute that comes with `Ploeh.AutoFixture.NSubstitute` does not. Wouldn't both of these scenarios fit in the 80%?

I completely agree that explicit test fixture configuration is better than implicit. That is why I'd prefer the usage scenarios listed _above_ instead of the following.

```C#
[Theory, MyTestConventions]
public void Test1(string isbn, IMyInterface iface);
```

The test fixture here is implicit. I think of it as a [standard fixture](http://xunitpatterns.com/Standard%20Fixture.html) implemented as a test conventions attribute. When used in teams with mixed expertise levels, _standard fixtures_ tend to devolve into [shared fixtures](http://xunitpatterns.com/Shared%20Fixture.html), which I prefer to avoid. By defining a global test convention attribute like the one above, I am discouraging the use of [minimal fixtures](http://xunitpatterns.com/Minimal%20Fixture.html) for individual tests and parameters (based on the `CustomizeAttribute` and attributes in the `DataAnnotations` namespace) and encouraging the use of a _standard fixture_. 

I hope we agree that _minimal_ and _standard_ fixture designs are both valid and have benefits making them uniquely suitable in different circumstances. Based on my interpretation, the AutoFixture design today supports both, by allowing users to implement _standard_ fixtures by subclassing the `AutoDataAttribute` and to implement _minimal_ fixtures by reusing a number of attributes from the `DataAnnotations` namespace and by subclassing the `CustomizateAttribute` when necessary.

The changes proposed in this pull request are an _incremental_ improvement of the _existing_ design. They enable implementing first-class support for attributes for minimal fixtures in other _AutoFixture_ assemblies and not just the core or the customer assemblies supported today.

I hope I was able to convince you that this pull request does not try to pull AutoFixture in a new direction. With that hope, could you address the _specific_ changes in this pull request? 
- Introduction of the `IParameterCustomizationProvider` interface in the `AutoFixture.Kernel`? (this is a key design change, but naming, definition, visibility and location of the new abstraction are flexible)
- Implementation of the `TestDataProvider` utility class in Kernel? (this is optional and can be easily eliminated)
- Changing of the `AutoDataAttribute` implementations to recognize the new `IParameterCustomizationProvider` interface? (this is key implementation change, but the implementation is flexible)
- Changing of the `AutoDataAttribute` implementations to use the `TestDataProvider` utility class? (this is also optional and can be removed)

ploeh commented:
What's already in AutoFixture is a result of a 6½ year journey (so far). Lots of things would be different if I'd known then what I know now.

I try to manage AutoFixture a bit like one would captain an oil tanker: if you want to stay clear of an obstacle, you have to start turning in good time before reaching it. While I'm sympathetic to the argument about established precedents, some common idioms have turned out to be less beneficial than others. If I think that a certain established way of doing things pull AutoFixture in the wrong direction, I'll stop it from spreading.

Adorning methods and arguments with attributes fall in a grey zone.

The main reason the attributes from `System.ComponentModel.DataAnnotations` just work out of the box is that these types are part of the Base Class Library, on which AutoFixture already has a dependency. In order to implement them, all that was needed was some concrete classes.

I have a much higher tolerance for adding concrete classes than I have for adding interfaces. Well-designed interfaces should represent *abstractions*, and these should be discovered instead of invented. The AutoFixture kernel already has too many poorly designed interfaces (some of which are fortunately already deprecated). I'm actively looking for ways to get rid of some of them. Adding new interfaces to the AutoFixture core library runs contrary to that effort.

If there's a way to do what you want that doesn't introduce new interfaces (or abstract base classes) it might be worth looking into.

Ultimately, I still don't find that this is much of an issue anyway. Some time back, someone did an analysis of what it would take to turn AutoFixture into a Portable Class Library, and one of the observations was that the data annotations are missing from some configurations. This might be the case for CoreCLR as well. In that case, my plan was to pull support for data annotations out of the core library, and add it to another library with extensions that you can pull in if you're running on the full platform. If this happens (it would be a breaking change), support for the `[RegularExpression]` attribute would be in exactly the same position as the `[Substitute]` attribute, and I would have no problem with that.

If you pull in types from a downstream dependency, I think it's only natural that you have to configure AutoFixture to tell it how to deal with that type. That's the current situation of `[Substitute]`.

If you pull in types from an upstream dependency, I wouldn't be surprised if AutoFixture just works with that type out of the box. That's the case of the data annotations.

olegsych commented:
Thank you for the response. I am disappointed that the prescribed solution for problems like #419 and #430 is for the users to know why they occur and how to configure AutoFixture and their test environment to fix them. While I realize that the answers may be obvious to someone who developed AutoFixture or spent enough time to understand how it works, the combination of the obscure errors with complexity of the core take too much value out of using it in the first place. With no productive way of addressing these issues in AutoFixture, it's hard to justify using it in projects I'm working on.

ploeh commented:
If AutoFixture errors are obscure, we can work on making them better.
